applications/DG-Max/Algorithms/DivDGMaxDiscretization.cpp

```diff
@@ -96,167 +142,87 @@ void DivDGMaxDiscretization<DIM>::initializeBasisFunctions(
-
-        stiffnessMatrix.resize(totalDoFs, totalDoFs);
-        stiffnessMatrix = elementIntegral.integrate(
-            (*it), [&](Base::PhysicalElement<DIM>& element) {
-                LinearAlgebra::MiddleSizeMatrix result, temp;
-                elementStiffnessMatrix(element, result);
-                elementScalarVectorCoupling(element, temp);
-                result += temp;
-                return result;
-            });
-
-        (*it)->setElementMatrix(stiffnessMatrix, ELEMENT_STIFFNESS_MATRIX_ID);
-
-        if (initialCondition) {
-            logger.assert_debug(
-                false, "Initial condition integration not ported yet.");
-        }
 
-        if (initialConditionDerivative) {
-            logger.assert_debug(
-                false,
-                "Initial condition derivative integration not ported yet.");
-        }
-
-        if (sourceTerm) {
-            sourceVector.resize(totalDoFs);
-            sourceVector = elementIntegral.integrate(
-                (*it), [&](Base::PhysicalElement<DIM>& element) {
-                    LinearAlgebra::MiddleSizeVector result;
-                    elementSourceVector(element, sourceTerm, result);
-                    return result;
-                });
-            (*it)->setElementVector(sourceVector, ELEMENT_SOURCE_VECTOR_ID);
+        for (const auto& elementVec : elementVectors) {
+            LinearAlgebra::MiddleSizeVector vec;
+            if (!elementVec.second) {
+                vec.resize(indexing.getNumberOfDoFs());
+            } else {
+                vec = elementIntegrator_.integrate(
+                    (*it), [&](Base::PhysicalElement<DIM>& element) {
+                        LinearAlgebra::MiddleSizeVector result;
+                        elementSourceVector(element, elementVec.second, result);
+                        return result;
+                    });
+            }
+            element->setElementVector(vec, elementVec.first);
         }
     }
 }
 
 template <std::size_t DIM>
 void DivDGMaxDiscretization<DIM>::computeFaceIntegrals(
     Base::MeshManipulator<DIM>& mesh,
-    DivDGMaxDiscretization<DIM>::FaceInputFunction boundaryCondition,
-    Stab stab) const {
-    LinearAlgebra::MiddleSizeMatrix faceMatrix(2, 2);
-    LinearAlgebra::MiddleSizeVector faceVector(2);
-    Integration::FaceIntegral<DIM> faceIntegral;
+    const std::map<std::size_t, FaceInputFunction>& boundaryVectors,
+    Stab stab) {
+    LinearAlgebra::MiddleSizeMatrix faceMatrix;
 
-    faceIntegral.setTransformation(
-        std::shared_ptr<Base::CoordinateTransformation<DIM>>(
-            new Base::HCurlConformingTransformation<DIM>()),
-        0);
-    faceIntegral.setTransformation(
-        std::shared_ptr<Base::CoordinateTransformation<DIM>>(
-            new Base::H1ConformingTransformation<DIM>()),
-        1);
+    Utilities::FaceLocalIndexing indexing;
+    indexing.reinit({0, 1});
 
     auto end = mesh.faceColEnd();
```

Comment:
could that not be in the for loop declaration

Reply:
What would the benefit be? Because that would constantly update the included unknowns every step instead of only once.

Reply:
Sorry, I meant the `auto end = mesh.faceColEnd();` could be inside the header of the for-loop

Reply:
It could, but in the past I found out that it was quite inefficient. Apparently the construction of `faceColEnd()` is quite expensive. I don't know the exact numbers (may be in a previous commit message), but it was at least a few percent maybe even more than 10%.

Reply:
Okay, I would maybe add a comment, otherwise that is sensible.

Reply:
Done